### PR TITLE
fix(server): omit absent request-log ids instead of logging them empty (#301)

### LINKED ISSUE #301: Session affinity keys on an empty session id, and is silent when it cannot key at all
## Scope clarification

The observed 9-judge-call TB-Lite failure was specifically **Harbor running Hermes** with `session_affinity = true` and `message_hash_fallback` omitted/false. Hermes did not send a native session header. Harbor did generate a stable UUID per trial attempt, but sent it as `proxy_x_session_id`, which Switchyard consumed for routing statistics but not affinity.

If `message_hash_fallback = true`, affinity was already best-effort functional by hashing the first user message. PR #308 therefore adds a fourth fix that recognizes Harbor's exact per-attempt ID: it makes Harbor/Hermes sticky without the fallback and supersedes the heuristic when the fallback is enabled. The empty-ID safeguard and diagnostic changes are general rather than Harbor/Hermes-specific.

Correction to section 1 below: current `main` already trims empty bare HTTP header values after #269. The empty-ID libsy safeguard still protects structured metadata containing an empty nested session value and direct callers that construct `Metadata { session_id: Some("") }`.

## Summary

Session affinity has two failure modes that are silent: it can key on an identity that
isn't one, and it can key on nothing at all. Both surface as a route that reports itself
as configured while behaving as if affinity were off -- or, in the first case, worse than
off. We hit the second while benchmarking TB-Lite and lost a while to it, then found the
first by reading the surrounding code.

## 1. An empty session id is treated as a real identity

`AffinityRouter::affinity_key` accepts whatever `metadata.session_id` holds:

```rust
// crates/libsy/src/algorithms/util/affinity.rs
if let Some(metadata) = request.metadata.as_ref()
    && let Some(session) = metadata.session_id.clone()
{
    ...
    return Some(AffinityKey::Session(session));
}
```

Header parsing does not normalise empty values — `resolve_path` in
`crates/protocol/src/metadata.rs` returns `Some(raw.clone())` verbatim — so a harness that
sends `x-switchyard-session-id:` with an empty value yields `Some("")`. Every request in
every task then shares `AffinityKey::Session("")`, the first task's model latches, and the
whole run inherits it. The message-hash fallback cannot save this: it lives in the `else`
branch that is now unreachable.

This is cross-task contamination that produces plausible-looking numbers, not an error.

The sibling consumer of the same field already guards it, which is what makes this look
like an oversight rather than intent:

```rust
// crates/libsy/src/algorithms/fall_through.rs
fn session_id(request: &Request) -> Option<String> {
    request.metadata.as_ref()?
        .session_id.as_deref()
        .filter(|id| !id.is_empty())   // affinity.rs has no equivalent
        .map(str::to_string)
}
```

`nonempty_header` in `crates/switchyard-server/src/routing_log.rs` applies the same guard.
No existing test covers an empty session id.

## 2. Affinity that can never key anything says nothing

`message_hash_fallback` defaults to `false` (`#[serde(default)]` on a `bool`). So this
configuration is silently inert for any harness that sends no session header:

```toml
session_affinity = true
# message_hash_fallback not set -> false
```

`affinity_key` returns `None` on every request, every turn is classified afresh, and there
is no warning anywhere — `warn!` appears nowhere in `affinity.rs` or in the server's
`config.rs`. The behaviour is documented in `docs/reference/toml_schema.md` and
`docs/routing_algorithms/llm_classifier_routing.md`, but nothing at runtime tells you it is
happening.

Concretely: `benchmark/run-baseline.sh` sends no session header, so any benchmark using
`session_affinity = true` without the fallback silently measures per-turn routing while
reporting a sticky configuration. We measured 9 judge calls across 9 turns of a single
task before spotting it.

## 3. Request logs cannot distinguish absent from empty

```rust
// crates/switchyard-server/src/lib.rs
session_id = self.session_id.as_deref().unwrap_or(""),
```

A missing session and an empty one both render as `session_id=`, which is precisely the
distinction you need when diagnosing either problem above.

### PR DESCRIPTION
**Pull request by chethanuk**
## What

- Drop `.unwrap_or("")` on `session_id`/`correlation_id` in `RequestLogContext::emit` and `emit_cancelled`, so an absent id omits the field instead of logging it as empty.
- Widen the test subscriber to capture event fields (not just the message) and add a table test over absent/empty/present id shapes.

## Why

The terminal request log renders `session_id=` for both "no session was sent" and "an empty session was sent" — an operator can't tell them apart. This is part 3 of #301; parts 1 and 2 landed in #308.

Closes NVIDIA-NeMo/Switchyard#301.

## Notes for reviewers

`requested_model`, `selected_model`, and `error` keep `unwrap_or("")` on purpose: `emit_cancelled` hardcodes `selected_model = ""`, so those three mean "absent implies empty" by design. The only non-test producer of `session_id` (`resolve_path`) never returns `Some("")` today, so the empty-string test row is a regression guard rather than a live bug — the user-visible change is that an absent session stops printing `session_id=`.

## Test plan

- `cargo test -p switchyard-server request_log` — new test fails today (`left: Some(""), right: None`), 4 passed after the fix
- `cargo fmt --all --check`
- `cargo clippy -p switchyard-server --all-targets -- -D warnings`
- `cargo test --workspace` — 586 passed, 0 failed, 1 ignored

`cargo clippy --workspace` fails on `crates/prefill-router` (`chunks_exact`), pre-existing and untouched here — reproduces on `origin/main` alone.

## Summary by CodeRabbit

- **Bug Fixes**
  - Request logs now omit session and correlation IDs when they are unavailable, instead of displaying empty values.
  - Logging consistently preserves the distinction between missing and empty identifiers for both completed and cancelled requests.

- **Documentation**
  - Added an unreleased changelog entry describing the updated request-log behavior.

### CODE REVIEW

File: crates/switchyard-server/src/lib.rs [hunk 1]
```diff
@@ -1908,9 +1909,28 @@ mod tests {
         let _ = tokio::time::timeout(Duration::from_secs(5), server).await;
     }
 
+    // One terminal request event: its level, its message, and every other field it
+    // recorded. A field the event never recorded is a missing key, not an empty value.
+    type CapturedEvent = (Level, String, BTreeMap<String, String>);
```
Comment (messiaen) [inline crates/switchyard-server/src/lib.rs:1914]: I would make this a struct instead of tuple.

**Comment by messiaen** [comment]
@nachiketb-nvidia can you review this one?

**Comment by afourniernv** [comment]
@chethanuk could you switch the capture type to a named struct as requested in the open review thread? @nachiketb-nvidia gentle bump for review once that is updated.
